File: php/2023/05.php
<?php

declare(strict_types=1);

require_once './tools.php';

$handle = fopen('input/05.txt', 'r');

final class Range {
    private readonly int $sourceEnd;
    private readonly int $destinationEnd;

    public function __construct(
        private readonly int $destinationStart,
        public readonly int $sourceStart,
        public readonly int $length,
    ) {
        $this->sourceEnd = $this->sourceStart + ($this->length - 1);
        $this->destinationEnd = $this->destinationStart + ($this->length - 1);
    }

    public function isSourceInRange(int $source): bool
    {
        return $source >= $this->sourceStart && $source <= $this->sourceEnd;
    }

    public function getDestination(int $source): int
    {
        $offset = $source - $this->sourceStart;

        return $this->destinationStart + $offset;
    }

    public function isDestinationInRange(int $destination): bool
    {
        return $destination >= $this->destinationStart && $destination <= $this->destinationEnd;
    }

    public function getSource(int $destination): int
    {
        $offset = $destination - $this->destinationStart;

        return $this->sourceStart + $offset;
    }
}

/**
 * @param array<Range> $map
 */
function findDestinationInRanges(array $map, int $source): int
{
    foreach ($map as $range) {
        if ($range->isSourceInRange($source)) {
            return $range->getDestination($source);
        }
    }

    return $source;
}

/** @var array<int> $seeds */
$seeds = [];

/** @var array<array{ranges: array<Range>, source_min: int, source_max: int}> $maps */
$maps = [
    'seed-to-soil' => [],
    'soil-to-fertilizer' => [],
    'fertilizer-to-water' => [],
    'water-to-light' => [],
    'light-to-temperature' => [],
    'temperature-to-humidity' => [],
    'humidity-to-location' => [],
];


while (($line = fgets($handle)) !== false) {
    if (str_starts_with($line, 'seeds:')) {
        $seeds = explode(' ', substr($line, 7));
        $seeds = array_map('intval', $seeds);

        continue;
    }

    if (trim($line) === '') {
        continue;
    }

    if (str_contains($line, ' map:')) {
        $mapName = trim(str_replace(' map:', '', $line));

        while (($_line = trim((string) fgets($handle))) !== '') {
            [$dest, $source, $length] = explode(' ', $_line);

            $maps[$mapName]['ranges'] ??= [];
            $maps[$mapName]['ranges'][] = new Range(
                (int) $dest,
                (int) $source,
                (int) $length,
            );
        }

        continue;
    }

    dd("error reading the file at line '$line'");
}

// now go over the maps to register the min and max source values
// so that we can first check these and not go through all ranges if it's not necessary
foreach ($maps as & $map) { // /!\ REFERENCE
    $map['source_min'] = PHP_INT_MAX;
    $map['source_max'] = -1;

    foreach ($map['ranges'] as $range) {
        $min = $range->sourceStart;
        $max = $range->sourceStart + $range->length;

        $map['source_min'] = min($map['source_min'], $min);
        $map['source_max'] = max($map['source_max'], $max);
    }
}
unset($map); // because of reference

startTimer();

$smallestLocation = -1;

foreach ($seeds as $seed) {
    $destination = $seed;
    foreach ($maps as $map) {
        $destination = findDestinationInRanges($map['ranges'], $destination);
    }

    if ($smallestLocation === -1 || $destination < $smallestLocation) {
        $smallestLocation = $destination;
    }
}

printDay("05.1: $smallestLocation"); // 0.55 ms

// --------------------------------------------------

rewind($handle);
startTimer();

// $smallestLocation = PHP_INT_MAX;
// $seedRanges = array_chunk($seeds, 2);
//
// foreach ($seedRanges as $i => $seedRange) {
//     $minSeed = $seedRange[0];
//     $length = $seedRange[1];
//     display("starting for seed range $i", $seedRange[0], $seedRange[1]);
//
//     $j = 0;
//     for (; $length >= 0; --$length) {
//         $destination = $minSeed + $length;
//         foreach ($maps as $mapName => $map) {
//             if ($destination < $map['source_min'] || $destination > $map['source_max']) {
//                 continue;
//                 // this "optimisation" doesn't work at all because the function findDestinationInRanges() still gets called as many times as before...
//             }
//
//             $destination = findDestinationInRanges($map['ranges'], $destination);
//         }
//
//         if ($destination < $smallestLocation) {
//             $smallestLocation = $destination;
//         }
//
//         $j++;
//         if ($j % 10_000 === 0) {
//             echo '.';
//         }
//         if ($j % 500_000 === 0) {
//             echo "length remaining: $length" . PHP_EOL;
//             exit;
//         }
//     }
// }
// Note : /!\ this doesn't works ! /!\
// is infinitely too slow, like 10 seconds to check a million seeds, and yet there is a total of a few billions seeds...


// All of this commented code above may be right but is way too slow.
// Instead of finding the location based on the seeds, we will try every location, starting from 0, until we find a matching seed.
// (I got the idea from Reddit)

$location = 0;
$seed = 0;

$reversedMaps = array_reverse($maps, true);

/**
 * @param array<Range> $map
 */
function findSourceInRanges(array $map, int $destination): int
{
    foreach ($map as $range) {
        if ($range->isDestinationInRange($destination)) {
            return $range->getSource($destination);
        }
    }

    return $destination;
}

$seedRanges = array_chunk($seeds, 2);

foreach ($seedRanges as & $seedRange) { // /!\ REFERENCE
    $seedRange[2] = $seedRange[0] + ($seedRange[1] - 1);
}
unset($seedRange);

do {
    $location++;

    $destination = $location;
    foreach ($reversedMaps as $mapName => $reversedMap) {
        $source = $destination;
        $destination = findSourceInRanges($reversedMap['ranges'], $destination);
        // display($mapName, $destination, $source);
    }
    // $destination is now a seed

    foreach ($seedRanges as $seedRange) {
        if ($destination >= $seedRange[0] && $destination <= $seedRange[2]) { // seed is within range
            $seed = $destination;
            // display($location, $seed);
            break(2);
        }
    }

    if ($location % 10_000 === 0) {
        echo '.';
    }
    if ($location % 1_000_000 === 0) {
        echo PHP_EOL;
    }
} while (true);

printDay("05.1: $location | seed ($seed)"); // 669 s / 11 m (location 37806486)

File: python/2023/05.py
import time

import utils

# Day 2023.05

f = open("inputs/05.txt", "r")
lines = f.readlines()

# test input
# lines: list[str] = """seeds: 79 14 55 13
#
# seed-to-soil map:
# 50 98 2
# 52 50 48
#
# soil-to-fertilizer map:
# 0 15 37
# 37 52 2
# 39 0 15
#
# fertilizer-to-water map:
# 49 53 8
# 0 11 42
# 42 0 7
# 57 7 4
#
# water-to-light map:
# 88 18 7
# 18 25 70
#
# light-to-temperature map:
# 45 77 23
# 81 45 19
# 68 64 13
#
# temperature-to-humidity map:
# 0 69 1
# 1 0 69
#
# humidity-to-location map:
# 60 56 37
# 56 93 4""".splitlines()


for i, line in enumerate(lines):
    lines[i] = line.replace("\n", "")

# --------------------------------------------------
# init

utils.startTimer()

def strListToIntList(line: str) -> list[int]:
    return list(map(int, line.split(' ')))

class RangeClass: # Range conflict with the range() global construct needed for part 2
    mapName: str
    destStart: int = 0
    sourceStart: int = 0
    length: int = 0

    def __init__(self, mapName: str, destStart: int, sourceStart: int, length: int):
        self.mapName = mapName
        self.destStart = destStart
        self.sourceStart = sourceStart
        self.length = length

    def isInRange(self, value: int) -> bool:
        return value >= self.sourceStart and value <= self.sourceStart + self.length - 1

    def getDestination(self, value: int) -> int:
        offset: int = value - self.sourceStart
        return self.destStart + offset

    # methods for part2
    def isInDestinationRange(self, value: int) -> bool:
        return value >= self.destStart and value <= self.destStart + self.length - 1

    def getSource(self, value: int) -> int:
        offset: int = value - self.destStart
        return self.sourceStart + offset

    def __str__(self):
        return f"Range [{self.mapName} dest={self.destStart} source={self.sourceStart} len={self.length}]"
# end class

seeds: list[int] = []
maps: dict[str, list[RangeClass]] = {}
mapName: str = ''

for line in lines:
    if line.startswith("seeds: "):
        seeds = strListToIntList(line.replace("seeds: ", ""))
        continue

    if line == '': continue

    if 'map:' in line:
        mapName: str = line.replace(' map:', '')
        maps[mapName] = []
        currentMap = maps[mapName]
        continue

    # the line is a range
    numbers = strListToIntList(line)
    maps[mapName].append(RangeClass(mapName, numbers[0], numbers[1], numbers[2]))

utils.endInit()


# --------------------------------------------------
# part 1

checkOrder: list[str] = [
    'seed-to-soil',
    'soil-to-fertilizer',
    'fertilizer-to-water',
    'water-to-light',
    'light-to-temperature',
    'temperature-to-humidity',
    'humidity-to-location',
]

utils.startTimer()

lowestLocation: int = -1
for seed in seeds:
    # get location for seed
    valueToCheck: int = seed

    for mapName in checkOrder:
        for rangeObject in maps[mapName]:
            if rangeObject.isInRange(valueToCheck):
                valueToCheck = rangeObject.getDestination(valueToCheck)
                break

    # by now valueToCheck must be a location

    if lowestLocation == -1:
        lowestLocation = valueToCheck
    elif valueToCheck < lowestLocation:
        lowestLocation = valueToCheck
# end for

utils.endPart('1', lowestLocation) # 1181555926 in 0.28ms


# --------------------------------------------------
# part 2

utils.startTimer()

# Same also as before but run a bazillion (probably) more times.
# It is much too slow, it has checked only 307 millions seeds in about one hour (85K per seconds)
# lowestLocation: int = 999_999_999_999
# checkedSeedCount: int = 0
# for i in range(0, int(len(seeds)  / 2 + 1), 2):
#     seedStart = seeds[i]
#     seedRange = seeds[i + 1]
#
#     for seed in range(seedStart, seedStart + seedRange):
#         checkedSeedCount += 1
#         # get location for seed
#         valueToCheck: int = seed
#
#         for mapName in checkOrder:
#             for rangeObject in maps[mapName]:
#                 if rangeObject.isInRange(valueToCheck):
#                     valueToCheck = rangeObject.getDestination(valueToCheck)
#                     break
#
#         # by now valueToCheck must be a location
#
#         if valueToCheck < lowestLocation:
#             lowestLocation = valueToCheck
#
#         if checkedSeedCount % 1_000_000 == 0:
#             print("checked seed count: ", checkedSeedCount, time.time())
# end for

# So instead of going from the seeds to the locations, we do the opposite.
# We check every location from 1 up to the first one that gives one of our starting seeds

checkOrder.reverse()
location: int = 37_800_000

while True:
    location += 1
    if location % 100_000 == 0:
        print(location, time.time())

    valueToCheck = location
    for mapName in checkOrder:
        for rangeObject in maps[mapName]:
            if rangeObject.isInDestinationRange(valueToCheck):
                valueToCheck = rangeObject.getSource(valueToCheck)
                break # go to next map

    # if we are here, we reached a seed,
    # but we still need to check that the seed is indeed within our seeds ranges
    seedRangeFound: bool = False
    for i in range(0, int(len(seeds)  / 2 + 1), 2):
        seedStart = seeds[i]
        seedLength = seeds[i + 1]

        if seedStart <= valueToCheck <= seedStart + seedLength - 1:
            seedRangeFound = True
            break

    if seedRangeFound:
        break
# end while

utils.endPart('2', location) # 37806486
# it took 577s or 9.6 minutes
# the PHP script took 669s/11m
